modules/lotamePanoramaIdSystem.js

```diff
@@ -0,0 +1,243 @@
+      cache.expiryTimestampMs = parseInt(rawExpiry, 0);
+    }
+  } catch (error) {
+    utils.logError(error);
+  }
+
+  return cache;
+}
+
+/**
+ * Clear a cached value from cookies and local storage
+ * @param {String} key
+ */
+function clearLotameCache(key) {
+  if (key) {
+    if (storage.cookiesAreEnabled()) {
+      let expirationDate = new Date(0).toUTCString();
+      storage.setCookie(key, '', expirationDate, 'Lax', undefined, undefined);
+    }
+    if (storage.hasLocalStorage()) {
+      storage.removeDataFromLocalStorage(key, undefined);
+    }
+  }
+}
+/** @type {Submodule} */
+export const lotamePanoramaIdSubmodule = {
+
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * Decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @returns {(Object|undefined)}
+   */
+  decode(value, configParams) {
+    return utils.isStr(value) ? { 'lotamePanoramaId': value } : undefined;
+  },
+
+  /**
+   * Retrieve the Lotame Panorama Id
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @param {ConsentData} [consentData]
+   * @param {(Object|undefined)} cacheIdObj
+   * @returns {IdResponse|undefined}
+   */
+  getId(configParams, consentData, cacheIdObj) {
+    let localCache = getLotameLocalCache();
+
+    let refreshNeeded = Date.now() > localCache.expiryTimestampMs;
+
+    if (!refreshNeeded) {
+      return {
+        id: localCache.data
+      };
+    }
+
+    const storedUserId = getProfileId();
+
+    const resolveIdFunction = function (callback) {
+      let queryParams = {};
+      if (storedUserId) {
+        queryParams.fp = storedUserId
+      }
+
+      if (consentData && utils.isBoolean(consentData.gdprApplies)) {
+        queryParams.gdpr_applies = consentData.gdprApplies;
+        if (consentData.gdprApplies) {
+          queryParams.gdpr_consent = consentData.consentString;
+        }
+      }
+      const url = utils.buildUrl({
+        protocol: 'https',
+        host: `id.crwdcntrl.net`,
+        pathname: '/id',
+        search: utils.isEmpty(queryParams) ? undefined : queryParams,
+      });
+      ajax(
+        url,
+        (response) => {
+          let responseObj = {};
+          if (response) {
+            try {
+              responseObj = JSON.parse(response);
+              saveLotameCache(KEY_EXPIRY, responseObj.expiry_ts);
+
+              if (utils.isStr(responseObj.core_id)) {
+                saveLotameCache(
+                  KEY_ID,
+                  responseObj.core_id,
+                  responseObj.expiry_ts
+                );
+              } else {
+                clearLotameCache(KEY_ID);
+              }
+
+              if (utils.isStr(responseObj.profile_id)) {
+                setProfileId(responseObj.profile_id);
+              } else {
+                clearLotameCache(KEY_PROFILE);
+                clearLotameCache(KEY_ID);
```

Comment:
maybe i'm misreading the code here...but this seems to say that if you had a situation where the response contained `core_id` but not `profile_id`, on line 210 you would write the `core_id` to cache, but then here you would immediately remove it from cache. is that expected behavior? if so, why not combine the statements to avoid writing and deleting the `core_id`.

Reply:
Good point. The server could send just a profile_id OR a profile_id and core_id but not just a core_id. I can rework to make that clearer

Reply:
Pushed up a fix